treasury-fees-monitor/data-collection/collect-fee-data.js

```diff
@@ -0,0 +1,183 @@
+    // eslint-disable-next-line no-await-in-loop
+    newBlock = await provider.getBlock(newBlockNumber);
+    newBlockTimestamp = newBlock.timestamp;
+
+    if (newBlockTimestamp > userTimestamp) {
+      // if the retrieved timestamp is higher than we expected, shift the high block down
+      highBlock = newBlock;
+      highBlockTimestamp = newBlockTimestamp;
+      highBlockNumber = newBlockNumber;
+    } else if (newBlockTimestamp < userTimestamp) {
+      // if the retrieved timestamp is lower than we expected, shift the high block down
+      lowBlock = newBlock;
+      lowBlockTimestamp = newBlockTimestamp;
+      lowBlockNumber = newBlockNumber;
+    } else {
+      // we found the timestamp exactly (this is unlikely, but we need to check)
+      return newBlock;
+    }
+  }
+
+  // the low block number should now have the highest timestamp that does not exceed the user
+  // requested timestamp
+  return lowBlock;
+}
+
+async function collectData(provider, filterInfo, timePeriod) {
+  const endBlockNum = await provider.getBlockNumber();
+  const endBlock = await provider.getBlock(endBlockNum);
+
+  const startTimestamp = (endBlock.timestamp - (timePeriod * SECONDS_IN_DAY));
+  const startBlock = await getBlockByTimestamp(provider, startTimestamp);
+  const startBlockNum = startBlock.number;
+
+  // chunk each getLogs call into a smaller block range so there is less of a chance that a
+  // 'too many logs' type error occurs
+  const promises = [];
+  for (let i = startBlockNum; i <= endBlockNum; i += BLOCK_CHUNK) {
+    const filterBlockStart = i;
+    const filterBlockEnd = (i + BLOCK_CHUNK > endBlockNum) ? endBlockNum : i + BLOCK_CHUNK;
+
+    // create the filter Object to look for this topic hash emitted from the specified address
+    const filter = {
+      fromBlock: filterBlockStart,
+      toBlock: filterBlockEnd,
+      topics: [filterInfo.topicHash],
+      address: filterInfo.address,
+    };
+
+    promises.push(provider.getLogs(filter));
+  }
+
+  return (await Promise.all(promises)).flat();
+}
+
+async function writeToCSV(format, dataSet, fileName) {
+  // iterate through each object in data and parse according to the format given
+  const formattedData = dataSet.map((data) => {
+    const dataToWrite = {};
+    data.log.args.forEach((arg, index) => {
+      // eslint-disable-next-line no-underscore-dangle
+      if (arg._isBigNumber) {
+        dataToWrite[format[index].id] = arg.toString();
+      } else {
+        dataToWrite[format[index].id] = arg;
+      }
+    });
+
+    dataToWrite.blockNumber = data.blockNumber;
+    dataToWrite.txHash = data.txHash;
+
+    return dataToWrite;
+  });
+
+  // write as a csv using the format provided
+  const csvWriter = createCsvWriter({
+    path: `${__dirname}/${fileName}.csv`,
+    header: format,
+  });
+
+  await csvWriter.writeRecords(formattedData);
+}
+
+(async () => {
+  const provider = new ethers.providers.JsonRpcProvider(process.env.JSON_RPC_URL);
```

Comment:
I don't know how much it would help with batching the requests, but it couldn't hurt.
```suggestion
  const provider = new ethers.providers.JsonRpcBatchProvider(process.env.JSON_RPC_URL);
```